### PR TITLE
Clarify surrogate elbos in `tracegraph_elbo`

### PR DESCRIPTION
**Pull request by mattwescott**
Adds comments on how `_compute_elbo_reparam` relates to the reference, and factors baseline construction out of `_compute_elbo_nonreparam`.

This is the second of three PRs from #1886

### CODE REVIEW

File: pyro/infer/tracegraph_elbo.py [hunk 1]
```diff
@@ -116,46 +171,20 @@ def _compute_elbo_non_reparam(guide_trace, non_reparam_nodes, downstream_costs):
...
-                raise ValueError("Expected baseline at site {} to be {} instead got {}".format(
-                    node, downstream_cost.shape, baseline.shape))
-            downstream_cost = downstream_cost - baseline
-        surrogate_elbo += (score_function_term * downstream_cost.detach()).sum()
+        score_function = guide_site["score_parts"].score_function
+
+        use_baseline, baseline_loss_term, baseline = _construct_baseline(node, guide_site, downstream_cost)
+
+        if use_baseline:
+            downstream_cost -= baseline
```
Comment (fritzo) [inline pyro/infer/tracegraph_elbo.py]: Please avoid introducing any new in-place ops.

Comment (mattwescott) [reply pyro/infer/tracegraph_elbo.py]: Fixed. After this batch of PRs, might also swap out the old in-place ops for #1893. Might as well while I'm neck deep in here?

File: pyro/infer/tracegraph_elbo.py [hunk 2]
```diff
@@ -92,20 +137,30 @@ def _compute_downstream_costs(model_trace, guide_trace,  #
 
 
 def _compute_elbo_reparam(model_trace, guide_trace):
+
+    # Compute a surrogate ELBO. In [1], this is simply the ELBO because
```
Comment (martinjankowiak) [inline pyro/infer/tracegraph_elbo.py]: i don't really understand this comment. in ref [1] they also make use of the language of a surrogate loss function (section 3.2)

File: pyro/infer/tracegraph_elbo.py [hunk 3]
```diff
@@ -92,20 +137,30 @@ def _compute_downstream_costs(model_trace, guide_trace,  #
...
 
-    # deal with log p(z|...) terms
+    # Bring log p(x, z|...) terms into both the ELBO and the surrogate
     for name, site in model_trace.nodes.items():
         if site["type"] == "sample":
             elbo += site["log_prob_sum"]
             surrogate_elbo += site["log_prob_sum"]
 
-    # deal with log q(z|...) terms
+    # Bring log q(z|...) terms into the ELBO, and effective terms into the
```
Comment (martinjankowiak) [inline pyro/infer/tracegraph_elbo.py]: what's an "effective term"?

File: pyro/infer/tracegraph_elbo.py (hunk 2 above)
Comment (mattwescott) [reply pyro/infer/tracegraph_elbo.py]: A more precise start to my comment might be: 

In ref [1], the portion of the surrogate loss handled by this function is `\sum{c}`, which in this context would simply be the ELBO...

File: pyro/infer/tracegraph_elbo.py (hunk 3 above)
Comment (mattwescott) [reply pyro/infer/tracegraph_elbo.py]: If reparameterizing a stochastic node inserted the transformation into the computation graph, say `z=t(epsilon, theta)`, the resulting cost terms `log q(epsilon|...)` naturally wouldn't contribute to the gradients. Here reparamerization is not explicit in the graph, instead by default it seems to be inferred from the attribute `has_rsample`. Then given that (non)reparameterization, the behavior under differentiation is stored in `score_parts`.

Coming from the paper, `site['score_parts'].entropy_term` felt like an effective term for `log q` given that the parameterization is implicit. 

Would it be useful to make these statements more precise?

File: pyro/infer/tracegraph_elbo.py (hunk 2 above)
Comment (mattwescott) [reply pyro/infer/tracegraph_elbo.py]: Thanks for engaging with this! Realized that if my image were right, the implementation would be wrong, and that has been a very useful signal to revisit my understanding. Would this be closer to the truth?

    # In ref [1], the part of the surrogate loss computed here is `\sum{c}`, 
    # which in this case is simply the ELBO. Depending on the parameterization, 
    # some of the entropy terms may not contribute (in expectation) to the 
    # gradient estimate. To reduce the variance of the gradient estimates under some
    # conditions, the log q(z|...) terms are instead taken from `site['score_parts']`

Comment (mattwescott) [reply pyro/infer/tracegraph_elbo.py]: @martinjankowiak marking these as (hopefully) resolved to simplify reviewing the new commit

**Comment by fritzo** [comment]
@martinjankowiak can you also approve? I'm happy to merge now.

**Comment by martinjankowiak** [comment]
great, thanks @mattwescott 

re: the 3rd PR you were considering, it'd be great if you outlined your rationale before getting started with any code.

some things that i think are worth keeping in mind:

- this is a complex piece of code, and so it can be a bit onerous to change. consequently i think it's best to leave it unchanged unless there are compelling reasons
- at some point in the new few months (?) PyTorch will have named tensor dimensions, and that will likely require more-or-less significant changes to this part of the codebase 
- at some point we plan to migrate to [funsor](https://github.com/pyro-ppl/funsor/), which will likely [deal](https://github.com/pyro-ppl/funsor/blob/416cff6dff9cec6f5a21d769edf90917e7c66bc0/funsor/torch.py#L283) with score function gradients rather differently

@fritzo thoughts?

**Comment by mattwescott** [comment]
Thanks @martinjankowiak. I'm going to try to bother you with one more commit from #1886, and I'll follow up on these leads and have some discussion before going further in that direction. 

> this is a complex piece of code, and so it can be a bit onerous to change. consequently i think it's best to leave it unchanged unless there are compelling reasons

For now my primary motivation is to understand this better, both from mucking around in the code and from your review. My hope is that the changes warrant the effort to review, even if they wouldn't by any means be a priority for the project.

> at some point in the new few months (?) PyTorch will have named tensor dimensions, and that will likely require more-or-less significant changes to this part of the codebase

Thanks for the heads up, was puzzling over some dimension logic in `TraceEnum_ELBO` just yesterday. 

> at some point we plan to migrate to funsor, which will likely deal with score function gradients rather differently

Very interesting!
